Server Side Field Validation Support For Kubectl

kevindelgado opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Modifies the behavior of the kubectl `--validate` flag to support server-side field validation.

Changes the kubectl --validate flag from a bool to a string that accepts the values {true, strict, warn, false, ignore}
* true/strict - perform validation and error the request on any invalid fields in the ojbect. It will attempt to perform server-side validation if it is enabled on the apiserver, otherwise it will fall back to client-side validation.
* warn - perform server-side validation and warn on any invalid fields (but ultimately let the request succeed by dropping any invalid fields from the object). If validation is not available on the server, perform no validation.
* false/ignore - perform no validation, silently dropping invalid fields from the object.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
This addresses the [kubectl flag](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2885-server-side-unknown-field-validation) portion of the beta graduation criteria of [KEP-2885](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2885-server-side-unknown-field-validation)

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Changes the kubectl --validate flag from a bool to a string that accepts the values {true, strict, warn, false, ignore}
* true/strict - perform validation and error the request on any invalid fields in the ojbect. It will attempt to perform server-side validation if it is enabled on the apiserver, otherwise it will fall back to client-side validation.
* warn - perform server-side validation and warn on any invalid fields (but ultimately let the request succeed by dropping any invalid fields from the object). If validation is not available on the server, perform no validation.
* false/ignore - perform no validation, silently dropping invalid fields from the object.

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2885-server-side-unknown-field-validation
```

kevindelgado commented:
/wip

kevindelgado commented:
I’ve manually tested kubectl create and apply, still figuring out how to write formal tests for this (and fix the ones that have broken), but wanted to get a review going in parallel on my initial ideas. 

The main files worth looking at first are:
* cmd/apply/apply.go
* cmd/create/create.go
* cmd/util/helpers.go
* validation/schema.go
* cmd/edit/editoptions.go
* factory_client_access.go
* cli-runtime/…/query_param_verifier.go

I think most of the others are just copypasta


/assign @apelisse
/cc @seans3 @liggitt

kevindelgado commented:
/retest

apelisse commented on staging/src/k8s.io/cli-runtime/pkg/resource/query_param_verifier.go:
```diff
@@ -0,0 +1,170 @@
+	"k8s.io/client-go/dynamic"
+)
+
+func NewQueryParamVerifier(dynamicClient dynamic.Interface, openAPIGetter discovery.OpenAPISchemaInterface) *QueryParamVerifier {
```
I think if you had done that, it has a couple of consequences on that PR:
1. You don't have to rename `dryRunVerifier` everywhere
2. You don't have to change "HasSupport" calls anywhere
3. You just need to change the instantiation of the dryRunVerifier into a QueryParamVerifier.

I guess the question is, how is the object most likely to be used. Also, I'm hoping the `openAPIGetter` is cached behind a `sync.Once` if we want to make sure we don't fetch the openapi multiple times.

From how I can see this class being used, I think it's fine if we have two different instances of a QueryParamVerifier, one for dryRun and one for validation.

liggitt replied:
Yeah, I agree. If you're wanting the instance to be shared so it can reuse a looked-up schema, you can have a looker-upper type that does lookup and takes a param as an arg, and another type that embeds that and a specific query param, then the dryRun and fieldValidation verifiers are two instances of that specific-param-support type that can share a looker-upper instance for schema caching

kevindelgado committed:
add server-side validation support to kubectl

kevindelgado commented on staging/src/k8s.io/cli-runtime/pkg/resource/query_param_verifier.go:
```diff
@@ -0,0 +1,170 @@
+	param VerifiableQueryParam
+}
+
+func NewParamUnsupportedError(gvk schema.GroupVersionKind, param VerifiableQueryParam) error {
```
done

kevindelgado replied:
now we are creating them outside the package for testing. Made it public again

kevindelgado commented on staging/src/k8s.io/cli-runtime/pkg/resource/query_param_verifier.go:
```diff
@@ -0,0 +1,170 @@
+	if err != nil {
+		// We assume that we couldn't find the type, then check for namespace:
+		supports, _ = supportsQueryParam(oapi, schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Namespace"}, param)
+		// If namespace supports dryRun, then we will support dryRun for CRDs only.
```
I updated the comment, but can you clarify what you mean by still correct? Is there an easy way to tell if it's still needed?

kevindelgado commented:
Okay, I've broken this change into the two commits as suggested, let me know if/when you want me to open up a separate PR to try to merge it first (or if you just want the commit separated but still reviewed in this PR).

Tried to address feedback around query param verifiers, see my [comment](https://github.com/kubernetes/kubernetes/pull/108350#discussion_r822832038)

PTAL @liggitt @apelisse

kevindelgado commented:
Okay, I’ve updated everything based on our discussion, reverting the looker-upper and expanded the query_param_verifier tests to incorporate field validation now. 

Had to update the fakeSchema swagger.json along the way.  

I’ve broken query param verifier into a separate PR at https://github.com/kubernetes/kubernetes/pull/108620 if we want to merge that first. (I’ll keep the two PR’s in sync though if you want to keep reviewing it here, but still merge the separate one first)

PTAL @apelisse @liggitt

kevindelgado commented:
I have now added the validation cmd tests. 

Notice that it only tests the `create` command, and with that only for creating deployments. 
I had originally envisioned a testing suite with dimensions of:

1. Create, update, all patch types (SMP, json-patch, json-merge-patch), ss-apply, cs-apply (for both create and update)
2. Builtins vs CRDs
3. SS-validation enabled vs disabled
4. Validate flag {strict, warn, ignore, true, false}

But 

* numbers 1 & 2 seem unnecessary. My reasoning is that this change really only affects the last dimension (validate flag), so even if we don’t have cmd tests for CRDs or all kinds of patches, those should be getting captured by the SS field validation integration tests.
* number 3 is really complicated given the test-cmd framework. Enabling feature-gates per test seems harder than it's worth

I spoke to Antoine about this fact that we can’t easily toggle the ss-validation enabled feature-gate, our conclusion was to write the test so that it passes against a disabled cluster, and then in a follow-up PR where we bump the feature to beta (and enable it by default) we will be forced to update the test so that it tests when ss-validation is enabled. This seems reasonable to me, but let us know what you think.

I've removed the WIP flag as this is the last code I have intended to add (pending review obviously). This test along with the `TestGetValidationDirective` unit test I added earlier are all the tests I was planning to add for this change, but let me know if there are more tests you think would be useful.

PTAL @apelisse @liggitt

kevindelgado commented:
/retest

kevindelgado commented:
Rebased now that https://github.com/kubernetes/kubernetes/pull/108620 has merged. It should be easier to review now. PTAL @apelisse @liggitt

kevindelgado commented on staging/src/k8s.io/kubectl/pkg/validation/schema.go:
```diff
@@ -101,3 +103,32 @@ func (c ConjunctiveSchema) ValidateBytes(data []byte) error {
+// being unsupported by the apiserver, because
+// server-side validation will be performed instead
+// of client-side validation.
+type ParamVerifyingSchema struct {
```
done

kevindelgado commented on staging/src/k8s.io/kubectl/pkg/validation/schema.go:
```diff
@@ -101,3 +103,32 @@ func (c ConjunctiveSchema) ValidateBytes(data []byte) error {
+}
+
+// ValidateBytes validates bytes per a ParamVerifyingSchema
+func (c ParamVerifyingSchema) ValidateBytes(data []byte) error {
```
Yea, I've added a unit test to specifically test the "HasSupport" logic. It doesn't seem appropriate to test the schemavalidation functions here because that is already captured by testing in the `schemavalidation` package.

As you can see, this test has also forced me to make the QueryParamVerifier an interface that we can mock out. I took the approach of only using the interface for where we explicitly need it for testing and left the rest of the uses in code to use the concrete `QueryParamVerifier` type still. Let me know if that's okay.

kevindelgado reviewed (commented):
All feedback addressed. PTAL @apelisse

kevindelgado commented on staging/src/k8s.io/kubectl/pkg/cmd/edit/edit.go:
```diff
@@ -90,7 +91,10 @@ func NewCmdEdit(f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobra
-	cmdutil.AddValidateOptionFlags(cmd, &o.ValidateOptions)
+	//cmdutil.AddValidateOptionFlags(cmd, &o.ValidateOptions)
+	// see above, why is this the only place that uses ValidateOptions?
+	// Can we use AddValidateFlags instead like we do everywhere else?
```
done

kevindelgado commented:
/retest

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -196,14 +196,20 @@ func (p *Patcher) Patch(current runtime.Object, modified []byte, source, namespa
 
+// Patch tries to patch an OpenAPI resource. On success, returns the merge patch as well
+// the final patched object. On failure, returns an error.
+func (p *Patcher) Patch(current runtime.Object, modified []byte, source, namespace, name string, errOut io.Writer) ([]byte, runtime.Object, error) {
```
are there still any callers to this we need to update?

liggitt replied:
> It seems like because there are so few callers it might be best to modify Patch to accept options, instead of having a separate PatchWithOptions. The diff caller can just past nil options like it currently does.

that seems better to me... then each caller gets to make a choice about what to pass in

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_deployment.go:
```diff
@@ -191,6 +197,7 @@ func (o *CreateDeploymentOptions) Run() error {
 		if o.FieldManager != "" {
 			createOptions.FieldManager = o.FieldManager
 		}
+		createOptions.FieldValidation = string(o.ValidationDirective)
```
was this file just done as an example? there look to be ~17 other create_*.go kubectl commands that follow the same pattern

kevindelgado replied:
yikes, good catch. No, definitely meant to be done for all of them, I think I just missed everything because I didn't understand what heuristic I could use to ensure that I had accounted for all of them. I think I have that now, which is that this is needed for every command that uses `cmd.AddValidateFlags`

I've gone through and added it now for all of them. Ideally, I imagine we would want some kind of test that ensured we added it to all of them (instead of relying on me to say "I've done it for every command that matches this heuristic", but what I can think of (adding a new cmd-test for every command) doesn't seem all that practical.

Let me know what you think.

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/util/factory_client_access.go:
```diff
@@ -141,8 +141,13 @@ func (f *factoryImpl) UnstructuredClientForMapping(mapping *meta.RESTMapping) (r
+func (f *factoryImpl) Validator(validationDirective ValidationDirective, verifier *resource.QueryParamVerifier) (validation.Schema, error) {
+	// client-side schema validation is only performed
+	// when the validationDirective is strict.
+	// If the directive is warn, we rely on the ParamVerifyingSchema
+	// to catch it and block the request, because we do not know
```
if the directive is warn, we shouldn't be blocking the request, right?

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:
```diff
@@ -397,11 +397,14 @@ func GetPodRunningTimeoutFlag(cmd *cobra.Command) (time.Duration, error) {
+		"strict",
+		`Must be "true", "false", "strict", "warn", or "ignore".
+		"true" or "strict" will use a schema to validate the input and fail the request if invalid. It will perform server side validation if ServerSideFieldValidation is enabled on the api-server, but will fall back to less reliable client-side validation if not.
+		"warn" will perform server-side field validation if it is enabled on the apiserver and will warn about any invalid fields but ultimately drop these fields from the object and let the request succeed. The request will fail with a 400 error if ServerSideFieldValidation is not enabled on the server.
```
> The request will fail with a 400 error if ServerSideFieldValidation is not enabled on the server

is that accurate? I would have expected the parameter to be ignored

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:
```diff
@@ -570,6 +573,39 @@ func GetFieldManagerFlag(cmd *cobra.Command) string {
+)
+
+func GetValidationDirective(cmd *cobra.Command) (ValidationDirective, error) {
+	var validateFlag = GetFlagString(cmd, "validate")
```
was `--validate` (with no value) valid and equivalent to `--validate=true` previously? if so, make sure we continue honoring that

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:
```diff
@@ -570,6 +573,39 @@ func GetFieldManagerFlag(cmd *cobra.Command) string {
+	}
+	// The flag was a boolean
+	if b {
+		// TODO: should we warn about deprecating the boolean value in favor of {strict, warn, ignore}?
```
I would not expect to warn. I can't think of any benefit to making people change --validate=true to --validate=strict and --validate=false to --validate=ignore

apelisse replied:
Yeah, it's because for dry-run we wanted to deprecate the boolean form, not sure we want to deprecate it here.

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:
```diff
@@ -397,11 +397,14 @@ func GetPodRunningTimeoutFlag(cmd *cobra.Command) (time.Duration, error) {
+	cmd.Flags().String(
+		"validate",
+		"strict",
+		`Must be "true", "false", "strict", "warn", or "ignore".
```
would it be clearer to group the equivalent values?
```suggestion
		`Must be one of: "strict" (or "true"), "warn", "ignore" (or "false").
```

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:
```diff
@@ -570,6 +573,39 @@ func GetFieldManagerFlag(cmd *cobra.Command) string {
+type ValidationDirective string
+
+const (
+	ValidationDirectiveStrict ValidationDirective = "Strict"
+	ValidationDirectiveWarn   ValidationDirective = "Warn"
+	ValidationDirectiveIgnore ValidationDirective = "Ignore"
```
is there a reason to re-declare these instead of using them from meta/v1?

liggitt commented on staging/src/k8s.io/kubectl/pkg/validation/schema.go:
```diff
@@ -101,3 +103,44 @@ func (c ConjunctiveSchema) ValidateBytes(data []byte) error {
+
+	err = c.verifier.HasSupport(gvk)
+	if resource.IsParamUnsupportedError(err) {
+		if c.directive == "Warn" {
+			return fmt.Errorf("cannot perform warn validation if fieldValidation parameter is unsupported")
+		}
```
we can emit a warning, but by definition should not fail the request

liggitt commented on staging/src/k8s.io/kubectl/pkg/validation/schema.go:
```diff
@@ -101,3 +103,44 @@ func (c ConjunctiveSchema) ValidateBytes(data []byte) error {
+
+	err = c.verifier.HasSupport(gvk)
+	if resource.IsParamUnsupportedError(err) {
+		if c.directive == "Warn" {
```
compare against the declared constant

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/validation/schema_test.go:
```diff
@@ -139,3 +142,97 @@ func TestConjunctiveSchema(t *testing.T) {
+
+func (v *mockVerifier) HasSupport(gvk schema.GroupVersionKind) error {
+	if !v.supported {
+		return resource.NewParamUnsupportedError(gvk, resource.QueryParamFieldValidation)
```
I think, there is no need to export that `NewParamUnsupportedError` function just because of this test. As far as I can see, you are checking error is nil or not. That would be better to return standard error for mocking?

ardaguclu commented on test/cmd/create.sh:
```diff
@@ -153,3 +153,63 @@ run_kubectl_create_kustomization_directory_tests() {
+  output_message=$(! kubectl create -f hack/testdata/invalid-deployment-unknown-and-duplicate-fields.yaml --validate=foo 2>&1)
+  kube::test::if_has_string "${output_message}" 'Invalid validate value'
+
+  set +o nounset
```
There should be a clean up for `create_and_use_new_namespace`. Deployment seems ok but for namespace which is created already?

kevindelgado replied:
oops, all these extra `create_and_use_new_namespace` calls are erroneous (copied from when I was duplicating test cases). I've removed them, so that there's only one `create_and_use_new_namespace` at the beginning of the test. I don't think we need to perform an explicit clean up though, because most tests seem to call it at the beginning of the test without a cleanup for it.

ardaguclu commented on test/cmd/create.sh:
```diff
@@ -153,3 +153,63 @@ run_kubectl_create_kustomization_directory_tests() {
 }
+
+# Runs tests related to kubectl create --validate
+run_kubectl_create_validate_tests() {
```
You might need to add this function in here https://github.com/kubernetes/kubernetes/blob/master/test/cmd/legacy-script.sh to be run.

kevindelgado replied:
thanks, done

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/util/editor/editoptions.go:
```diff
@@ -561,8 +574,11 @@ func (o *EditOptions) annotationPatch(update *resource.Info) error {
+	}
 	helper := resource.NewHelper(client, mapping).WithFieldManager(o.FieldManager)
-	_, err = helper.Patch(o.CmdNamespace, update.Name, patchType, patch, nil)
+	_, err = helper.Patch(o.CmdNamespace, update.Name, patchType, patch, patchOptions)
```
Are we going to use `Patch` or `PatchWithOptions`?

kevindelgado reviewed (commented):
@liggitt addressed all of your feedback. Broke it up by commit into the three main areas:
1) using meta/v1 constats
2) not failing warn requests
3) FieldValidation request options

PTAL

kevindelgado reviewed (commented):
all feedback addressed @ardaguclu

kevindelgado commented:
gentle ping @liggitt

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -566,7 +575,10 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
 		if o.DryRunStrategy != cmdutil.DryRunClient {
 			// Then create the resource and skip the three-way merge
-			obj, err := helper.Create(info.Namespace, true, info.Object)
+			createOptions := &metav1.CreateOptions{
+				FieldValidation: o.ValidationDirective,
+			}
+			obj, err := helper.CreateWithOptions(info.Namespace, true, info.Object, createOptions)
```
Why does the helper need the validation directive twice?

kevindelgado replied:
ya, you're totally right, this is unecessary now that we've added `WithFieldValidation` to the helper. I didn't realize that we're using the helper here (and elsewhere). I've reverted all of it in the most recent commit.

apelisse commented:
Also, can you clarify the behavior of "warn" when the feature is disabled by changing the body of this PR (including the release notes?)

apelisse commented:
OK I think I'm happy with it, feel free to squash some of these commits. I'll leave this here for Jordan to also chime-in.

kevindelgado commented:
/retest

kevindelgado commented:
/retest

fedebongio commented:
/triage accepted

kevindelgado commented:
/retest

liggitt commented:
> This PR however, is ready for a review from an approver

ack, on my short list, I should be able to get to it tonight

kevindelgado commented:
/retest

apelisse reviewed (commented):
OK i've made another pass, I think it's good, thanks!

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:
```diff
@@ -570,6 +573,30 @@ func GetFieldManagerFlag(cmd *cobra.Command) string {
+		case "ignore":
+			return metav1.FieldValidationIgnore, nil
+		default:
+			return metav1.FieldValidationStrict, fmt.Errorf(`Invalid validate value (%v). Must be "true", "false", "strict", "warn", or "ignore".`, validateFlag)
```
nit: take it or leave it:
```suggestion
			return metav1.FieldValidationStrict, fmt.Errorf(`invalid --validate option %q; must be one of: strict (or true), warn, ignore (or false)`, validateFlag)
```

liggitt commented:
changes in this PR lgtm. before merge, we need to understand any test adjustments that need to be made once kubectl starts using server-side validation.

in https://github.com/kubernetes/kubernetes/pull/108889, stack the following commits (one commit for each item):
1. this PR
2. enabling the feature gate
3. adjusting existing tests to pass with or without server-side field validation
4. any new tests specific to server-side validation we want to exercise

Once #108889 is green, pick commit 3 to this PR so we can make _sure_ those adjustments work with/without server-side field validation enabled.

I expect once we merge #108889, we'll also have to pick that commit 3 to release-1.23 to make kubectl skew tests work

kevindelgado commented:
/test pull-kubernetes-e2e-kind-ipv6

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:
```diff
@@ -397,11 +397,14 @@ func GetPodRunningTimeoutFlag(cmd *cobra.Command) (time.Duration, error) {
+		"strict",
+		`Must be one of: strict (or true), warn, ignore (or false).
+		"true" or "strict" will use a schema to validate the input and fail the request if invalid. It will perform server side validation if ServerSideFieldValidation is enabled on the api-server, but will fall back to less reliable client-side validation if not.
+		"warn" will perform server-side field validation if it is enabled on the apiserver and will warn about any invalid fields but ultimately drop these fields from the object and let the request succeed. This flag is ignored (and we default to ignoring) if server-side field validation is not supported by the API server.
```
"we default to ignoring" is confusing, since we actually default to `strict`.

```suggestion
		"warn" will warn about unknown or duplicate fields without blocking the request if server-side field validation is enabled on the API server, and behave as "ignore" otherwise.
```

kevindelgado commented:
ok, looks like everything is green here now (sans broken "pull-kubernetes-integration")

liggitt commented:
> ok, looks like everything is green here now (sans broken "pull-kubernetes-integration")

but the break is related to the changes in this PR, right? (run_kubectl_create_validate_tests is the failure)

kevindelgado commented:
> > ok, looks like everything is green here now (sans broken "pull-kubernetes-integration")
> 
> but the break is related to the changes in this PR, right? (run_kubectl_create_validate_tests is the failure)

ya, good catch. Sorry this was a silly mistake when I changed the error message on the invalid option. Fixing it now give me two minutes

kevindelgado commented:
ok fixed now, will update once CI runs

liggitt commented:
once CI is green here, can rebase https://github.com/kubernetes/kubernetes/pull/108889 on this commit

kevindelgado committed:
Expand newly created cmd-test

Modify newly created cmd-test to work for both
client-side and server-side validation

kevindelgado committed:
Adjust validation checks to pass for both client-side and server-side validation

kevindelgado commented:
Ok, I have now applied the test expansion commit from #108889 to this PR. In order to also open a PR against `release-1.23` I had to split the commit into two since part of it was only relevant to a test first created by this PR (hence why you see three commits here).

PR against release-1.23 is up at https://github.com/kubernetes/kubernetes/pull/109041 and #108889 is up to date and rebased on top of this PR

kevindelgado commented:
Everything is green across https://github.com/kubernetes/kubernetes/pull/108350 https://github.com/kubernetes/kubernetes/pull/108889 https://github.com/kubernetes/kubernetes/pull/109041 
PTAL @liggitt

kevindelgado commented:
All feedback addressed and PRs synced

liggitt commented:
/lgtm
/approve

liggitt commented:
/hold

will merge as part of https://github.com/kubernetes/kubernetes/pull/108889

liggitt commented:
/retest
/hold cancel

if this merges before that gets the conformance description acked, that's fine
